localization/ndt_scan_matcher/src/pose_array_interpolator.cpp

```diff
@@ -0,0 +1,120 @@
+  new_pose_ptr_ = std::make_shared<PoseWithCovarianceStamped>();
+
+  interpolate(pose_msg_ptr_array, target_ros_time);
+
+  // check the time stamp
+  validate_time_stamp_difference(old_pose_ptr_->header.stamp, target_ros_time, pose_timeout_sec);
+  validate_time_stamp_difference(new_pose_ptr_->header.stamp, target_ros_time, pose_timeout_sec);
+
+  // check the position jumping (ex. immediately after the initial pose estimation)
+  validate_position_difference(
+    old_pose_ptr_->pose.pose.position, new_pose_ptr_->pose.pose.position,
+    pose_distance_tolerance_meters);
+
+  // all validations must be true
+  if (!success_) {
+    RCLCPP_WARN(logger_, "Validation error.");
+  }
+}
+
+PoseArrayInterpolator::PoseArrayInterpolator(
+  rclcpp::Node * node, const rclcpp::Time target_ros_time,
+  const std::deque<PoseWithCovarianceStamped::ConstSharedPtr> & pose_msg_ptr_array)
+: logger_(node->get_logger()), clock_(*node->get_clock())
+{
+  success_ = true;
+  current_pose_ptr_ = std::make_shared<PoseWithCovarianceStamped>();
+  old_pose_ptr_ = std::make_shared<PoseWithCovarianceStamped>();
+  new_pose_ptr_ = std::make_shared<PoseWithCovarianceStamped>();
+
+  interpolate(pose_msg_ptr_array, target_ros_time);
+}
+
+void PoseArrayInterpolator::interpolate(
+  const std::deque<PoseWithCovarianceStamped::ConstSharedPtr> & pose_msg_ptr_array,
+  const rclcpp::Time target_ros_time)
+{
+  if (pose_msg_ptr_array.size() <= 1) {
+    RCLCPP_WARN_STREAM_THROTTLE(logger_, clock_, 1, "No Pose!");
+    success_ = false;
+    return;
+  }
+  get_nearest_timestamp_pose(pose_msg_ptr_array, target_ros_time, old_pose_ptr_, new_pose_ptr_);
+  const geometry_msgs::msg::PoseStamped interpolated_pose_msg =
+    interpolate_pose(*old_pose_ptr_, *new_pose_ptr_, target_ros_time);
+  current_pose_ptr_->header = interpolated_pose_msg.header;
+  current_pose_ptr_->pose.pose = interpolated_pose_msg.pose;
+  current_pose_ptr_->pose.covariance = old_pose_ptr_->pose.covariance;
+}
+
+geometry_msgs::msg::PoseWithCovarianceStamped PoseArrayInterpolator::get_current_pose()
+{
+  return *current_pose_ptr_;
+}
+
+geometry_msgs::msg::PoseWithCovarianceStamped PoseArrayInterpolator::get_old_pose()
+{
+  return *old_pose_ptr_;
+}
+
+geometry_msgs::msg::PoseWithCovarianceStamped PoseArrayInterpolator::get_new_pose()
+{
+  return *new_pose_ptr_;
+}
+
+bool PoseArrayInterpolator::is_success() { return success_; }
+
+void PoseArrayInterpolator::validate_time_stamp_difference(
+  const rclcpp::Time & target_time, const rclcpp::Time & reference_time,
+  const double time_tolerance_sec)
+{
+  const double dt = std::abs((target_time - reference_time).seconds());
+  if (dt > time_tolerance_sec) {
+    RCLCPP_WARN(
+      logger_,
+      "Validation error. The reference time is %lf[sec], but the target time is %lf[sec]. The "
+      "difference is %lf[sec] (the tolerance is %lf[sec]).",
+      reference_time.seconds(), target_time.seconds(), dt, time_tolerance_sec);
+    success_ = false;
```

Comment:
This function changes the internal state. 
But in terms of referential transparency, this function should explicitly return a boolean value and change the state in the caller. By doing so, this function can be const.

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/1757/commits/0e313563d9db70149f80da3523e4794a143c2bf8
resolved

Reply:
Yes
The equal case does not make a difference here.